Notebook 103

lkriener opened the pull request.

lkriener committed:
rename to something without spaces

lkriener committed:
Adapt section 103

- remove lambda funcs
- add some explanations
- remove uni HD specific stuff

timomit commented:
Bei mir funktioniert dieses nose nicht. Und ich check auch gar nicht, warum man nose hier pberhaupt braucht und das nicht einfach mit assert macht. 

Ich hab jedenfalls in 101 und 102 alle nose-imports rausgemacht und durch normale standard-asserts ersetzt

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -69,9 +69,51 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 2,
```
Und hast du nb-stripout angewendet? Eigentlich sollten die ganzen Metadaten und Zellen-Ausgaben entfernt sein... :)

lkriener replied:
ja hab ich verwendet...

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -299,31 +409,6 @@
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "### Einzeilige `lambda`-Funktionen sind praktisch für Mathematik\n",
```
Willst du Lambda-Funktionen nicht behandeln? :thinking: 
Wahrscheinlich hast du Recht. Ich persönlich benutze sie auch mega selten eigentlich...

timomit commented:
Ich würde fast sagen, dass dieser Abschnitt https://github.com/lkriener/dsa-python-course-1/blob/511b6c751812dea954882ecc3f933ba9b1342525/103_Funktionen_und_Module.ipynb#L416 hier noch etwas ausführlicher sein könnte. Das mit den Scopes ist vllt ein Konzept, was einem Anfänger nicht ganz klar ist. 
Außerdem sollte man irgendwo erwähnen, dass globale Variablen während der gesamten Laufzeit der Kernels erhalten bleiben, selbst wenn man den Variablennamen aus allen Codezellen löscht! (kann vllt am Anfang zu verwirrung und komischem Verhalten führen, wenn man das nicht checkt!)

timomit commented:
Vllt sollte man https://github.com/lkriener/dsa-python-course-1/blob/511b6c751812dea954882ecc3f933ba9b1342525/103_Funktionen_und_Module.ipynb#L536 hier ergänzen, dass die Punkt-Notation für zwei unterschiedliche Konzepte verwendet wird: Methoden in der OOP und namespaces bzw. module. Damit das nicht für verwirrung sorgt!

timomit commented:
Ich würde zum Schluss noch mehr Übungsaufgaben einfügen. Man könnte vllt die Aufgaben aus 102 recyclen und die engineering-Aufgaben mit Funktionen übersichtlicher machen!

lkriener committed:
replace nose with asserts for checking exercises

lkriener committed:
Improve 103

- mehr Aufgaben
- bessere Erklärung zu globalen Variablen
- Hinweis zu module.function() vs object.method()

lkriener committed:
call npstripout manually

lkriener commented:
> Ich würde zum Schluss noch mehr Übungsaufgaben einfügen. Man könnte vllt die Aufgaben aus 102 recyclen und die engineering-Aufgaben mit Funktionen übersichtlicher machen!

Ich hab unterwegs jetzt noch andere Aufgaben eingefügt. Bei den globalen Variablen. Erschien mir wichtiger und ich wollte es nicht überladen

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -464,9 +490,67 @@
+   "source": [
+    "### Aufgabe 3: Globale Variablen\n",
+    "\n",
+    "**a)** Überlege dir, ob du genau erklären kannst, warum welcher Wert von `GLOBAL_VAR` wo geprinted wird.\n",
```
Ich find die Aufgabe mega gut! :) Ich würde nur aufpassen: 1.) muss man wahrscheinlich den Kernel immer wieder resetten, sonst passieren evtl- komische Dinge, weil Variablen und Funktionen schon definiert sind und 2.) Ich würde vllt mit kommentaren im code hinweisen, welche Stelle du genau in welcher Teilaufgaben meinst. Damit es zu keiner Verwirrung kommt.

lkriener replied:
1.) darauf hatte ich eigentlich geachtet und gedacht, dass so wie es gecodet ist, nicht der Fall sein sollte...bist du anderer Meinung?

timomit replied:
Ja, also wenn man genau das macht, was du beschreibst, ist das so! Aber wenn die TN versehentich was anderes machen (z.B. weil sie die Aufgabe missverstehen), könnten komische Dinge passieren, die sie sich nicht erklären können. (Vllt...)

lkriener replied:
2.) gute idee, ich lass mir was einfallen

timomit replied:
Anyway. Ich bin sowieso der Meinung, dass man irgendwo im Tutorial erwähnen sollte, dass Jupyter einen internen State hat, der Variablen speichert, auch wenn die in allen Codeblöcken gelöscht sind. --> Das könnte auch für verwirrung sorgen, insb. bei Leuten, die schon coden können aber bislang nur Scripts ausgeführt haben, die nach ihrer Laufzeit beendet werden... Was meinst du?

timomit replied:
Ich weiß, es gibt sooo viel zu erwähnen, und wir wollen das Tutorial nicht zuuuu voll packen! Aber das ist vllt wirklich was, was einem graue Haare bereiten könnte...

lkriener replied:
1.) nevermind. Wenn man den code der Teilaufgaben überprüft, dann wird es zum Problem. Daran hatte ich nicht gedacht. Hab nur die mehrfache Ausführung von der oberen Zelle bedacht. Ich schreib ne Notiz dazu.

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -506,6 +590,63 @@
+    "\n",
+    "Bei der Verwendung von globalen Variablen ist Vorsicht geboten. Je länger der Code, desto schwieriger ist es den kompletten Überblick zu behalten, was welche Variable wo tut und wo sie wie verändert wird. \n",
+    "\n",
+    "### Aufgabe 4: Noch mehr globale Variablen\n",
```
Diese Aufgabe find ich klasse! :+1:

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -594,6 +735,13 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> **Hinweis:** Diese Notation mit `name.funktions_name()` hast du auch schon bei den Objekten und deren Methoden gesehen (z.B. `dict.keys()`). Das sieht zwar gleich aus wie die Notation mit den Modulen, aber ist es ist nicht genau das Gleiche...Hier wird die Notation benutzt um klarzumachen aus welchem Modul wir eine Funktion benutzen. Also nicht verwirren lassen."
```
Vielleicht vielleicht vielleicht kann man hier namespaces erwähnen: Irgendwie sowas à la: "Oft enthalten zwei unterschiedliche Module Funktionen mit dem gleichen Namen, wie z.B. math.exp() und np.exp(). Damit Python weiß, welche Implementierung benutzt werden soll, müssen wir das Modul mit der Punktnotation spezifizieren. Dieses Konzept wird auch Namespaces genannt."

Overloading funktioniert übrigens auch: Wenn du eine eigene Funktion schreibst, die eln() heißt, "überschreibst" du damit evtl. schon eine existierende Python Funktion! Pass also auf, wie du deine Funktionen oder Vaiablen benennst! 
Der Editor (oder Jupyter) können dir dabei helfen: Standard-Pythongefehle werden meistens anders gefärbt dargestellt. Wenn du also versehentlich eine Funktion oder Variable wie ein schon existierender Pythonbefehl benennst, wirst du das an seiner Farbe erkennen! Dann solltest du dir am besten einen anderen Namen überlegen!

timomit reviewed (commented):
ah, ich hab tatsächlich vergessen, hier nochmals auf "submit review" zu klicken! :see_no_evil: --> Wieder was dazugelernt!

lkriener committed:
incorporate timos comments

timomit commented on 103_Funktionen_und_Module.ipynb:
````diff
@@ -534,6 +555,10 @@
     "m.function_name()\n",
     "```\n",
     "\n",
+    "> Hinweis: Manchmal enthalten zwei unterschiedliche Module Funktionen mit dem gleichen Namen, wie z.B. math.exp() und np.exp(). Damit Python weiß, welche Implementierung benutzt werden soll, müssen wir das Modul mit der Punktnotation spezifizieren. Dieses Konzept wird auch Namespaces genannt.\n",
````
Da sind noch kleine Rechtschreibfehler drin (i.e. Pythongefehle)...

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -402,13 +410,26 @@
+    "- Klicke dann im Sub-Menü auf \"Restart Kernel...\"\n",
     "\n",
-    "**d)** Überlege dir, was passieren würde, wenn wir die Funktionsdefinition  `def print_global_var():` nach `GLOBAL_VAR = 17` verschieben würden.\n",
+    "Der Grund dafür ist, dass sich Jupyter Notebooks ein bisschen anders verhalten als normale Python Skripte. Wenn ein Python Skript einmal ausgeführt wird und dann an seinem Ende ankommt, wird es geschlossen und alle Variablen und deren Werte etc werden vergessen. Von einer Ausführung des Skripts zur nächsten wird nichts übertragen. Bei einer Jupyter Notebook Zelle ist das anders: Die Variable ...
```
Gut!

timomit commented on 103_Funktionen_und_Module.ipynb:
```diff
@@ -402,13 +410,26 @@
     "**a)** Überlege dir, ob du genau erklären kannst, warum welcher Wert von `GLOBAL_VAR` wo geprinted wird.\n",
     "\n",
-    "**b)** Überlege dir, was passieren würde, wenn wir die Funktionsdefinition  `def print_global_var():` ganz nach oben in der Zelle verschieben würden.\n",
+    "**b)** Überlege dir, was passieren würde, wenn wir die Funktionsdefinition  `def print_global_var():` ganz nach oben in der Zelle verschieben würden (Location b).\n",
```
top!

timomit reviewed (commented):
Find ich gut! Hab nur noch kleine Rechtschreibfehler gefunden...

lkriener committed:
fix typos
